### ISSUE TITLE
Add allowPattern to no-invalid-this

### ISSUE THREAD

**Issue by halleycarleton** [eslint/eslint#7036]
**What version are you using?**
3.4.0

**What did you do?**
Wrote mocha tests with `no-invalid-this` enabled

```js
/* eslint-env node, mocha */
/* eslint no-invalid-this: "error" */
describe(function() {
  beforeEach(function() {
    this.original = obj.prop;
  });
  afterEach(function() {
    obj.prop = this.original;
  });
});
```

**What happened?**
Received error `(no-invalid-this) Unexpected 'this'.`

**What did you expect to happen?**
Expected not to receive errors for using `this` keyword in `describe` callback when env is mocha.

I think the best solution would be to add an allowPattern option(or similar) to `no-invalid-this` to allow the user to define functions where the callback should have access to `this`.

**Comment by kaicataldo** [comment]
Hey there, thanks for the issue! I'm interested in a more solid proposal of how this new option would work, since the rule is behaving as intended.

**Comment by eslintbot** [comment]
Hi @halleycarleton, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by halleycarleton** [comment]
@eslintbot I actually already followed the link to Proposing a Rule Change and filled out the appropriate information.

@kaicataldo Sure, I can spend some time this weekend putting together something a bit more concrete. The rule is certainly behaving as intended, this is more of a request for an added feature to the rule.

**Comment by kaicataldo** [comment]
@halleycarleton Thanks, that would be great! I'm curious what you envision the config would look like and what kind of check `allowPattern` would make to know whether a pattern is allowed or not.

**Comment by mysticatea** [comment]
There was related discussion: #3600

**Comment by mysticatea** [comment]
Ah, also there was another discussion at the first time: https://github.com/eslint/eslint/pull/2857#issuecomment-122553664

**Comment by halleycarleton** [comment]
@mysticatea Thanks for pointing out those previous discussions. Since this has already been discussed I will close this issue.
